runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,277 @@
+	unsigned::TransactionValidityError,
+};
+use orml_asset_registry::AssetMetadata;
+use orml_traits::MultiCurrency;
+use pallet_transaction_payment::OnChargeTransaction;
+use pallet_xcmp_handler::XcmCurrencyData;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Get, PostDispatchInfoOf, Saturating, Zero},
+	transaction_validity::InvalidTransaction,
+	SaturatedConversion,
+};
+use sp_std::marker::PhantomData;
+
+#[derive(Debug)]
+pub struct FeeInformation {
+	token_id: TokenId,
+	xcm_data: Option<XcmCurrencyData>,
+	asset_metadata: Option<AssetMetadata<Balance, CustomMetadata>>,
+}
+pub trait CallParser<Call> {
+	fn fee_information(call: &Call) -> FeeInformation;
+}
+pub struct FeeCallParser;
+impl CallParser<Call> for FeeCallParser {
+	fn fee_information(c: &Call) -> FeeInformation {
+		if let Call::AutomationTime(pallet_automation_time::Call::schedule_xcmp_task {
+			para_id,
+			currency_id,
+			..
+		}) = c.clone()
+		{
+			let xcm_data = XcmpHandler::get_xcm_chain_data(u32::from(para_id), currency_id);
+			let asset_metadata = AssetRegistry::metadata(currency_id);
+
+			FeeInformation { token_id: currency_id, xcm_data, asset_metadata }
+		} else {
+			FeeInformation { token_id: NATIVE_TOKEN_ID, xcm_data: None, asset_metadata: None }
+		}
+	}
+}
+
+pub struct NativeFeeProcessor<T, C>(PhantomData<(T, C)>);
+impl<T, C> NativeFeeProcessor<T, C>
+where
+	T: pallet_transaction_payment::Config,
+	C: Currency<<T as frame_system::Config>::AccountId>,
+{
+	fn withdraw_fee(
+		who: &T::AccountId,
+		fee: C::Balance,
+		tip: C::Balance,
+	) -> Result<Option<NegativeImbalanceOf<C, T>>, TransactionValidityError> {
+		let withdraw_reason = if tip.is_zero() {
+			WithdrawReasons::TRANSACTION_PAYMENT
+		} else {
+			WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+		};
+
+		match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+			Ok(imbalance) => Ok(Some(imbalance)),
+			Err(_) => Err(InvalidTransaction::Payment.into()),
+		}
+	}
+}
+
+pub struct ForeignFeeProcessor<T, MC, TA>(PhantomData<(T, MC, TA)>);
+impl<T, MC, TA> ForeignFeeProcessor<T, MC, TA>
+where
+	T: pallet_transaction_payment::Config,
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	MC::CurrencyId: From<TokenId>,
+	TA: Get<T::AccountId>,
+{
+	fn withdraw_fee(
+		who: &T::AccountId,
+		fee: MC::Balance,
+		call_information: FeeInformation,
+	) -> Result<(), TransactionValidityError> {
+		if call_information.xcm_data.is_none() || call_information.asset_metadata.is_none() {
+			return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+		}
+
+		let currency_id = call_information.token_id.into();
+		let foreign_fee: MC::Balance = call_information
+			.asset_metadata
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.convert_fee_into_foreign(fee.saturated_into())
+			.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+			.saturated_into();
```

Comment:
General question- does this code need to guard against the fee being `0`? If the fee is `0` does that imply that there was a math bug or misconfiguration? Basically, does a fee of `0` here present a security risk to our chain and should we explicitly deny it?